src/pairs/mento-v2.ts

```diff
@@ -0,0 +1,255 @@
+    const [tokenInBucketSize, tokenOutBucketSize] =
+      this.matchBuckets(inputToken);
+
+    const amountOut = getAmountOut(
+      tokenInBucketSize,
+      tokenOutBucketSize,
+      spread,
+      inputAmount
+    );
+    return amountOut;
+  }
+
+  public snapshot(): PairMentoV2Snapshot {
+    return {
+      swapFee: this.swapFee.toFixed(),
+      pricingModule: this.pricingModule,
+      bucket0: this.bucket0.toPrecision(),
+      bucket1: this.bucket1.toPrecision(),
+    };
+  }
+
+  public async restore(snapshot: PairMentoV2Snapshot): Promise<void> {
+    this.swapFee = new BigNumber(snapshot.swapFee);
+    this.pricingModule = snapshot.pricingModule;
+    this.bucket0 = new BigNumber(snapshot.bucket0);
+    this.bucket1 = new BigNumber(snapshot.bucket1);
+  }
+
+  private mentoBucketsAfterUpdate = async () => {
+    /*
+    Just like in mento.ts but use the PoolExchange in order to read:
+    lastBucketUpdate, updateFrequency and spread
+    also the bucket0 and bucket1.
+    Refreshed the PoolExchange structure first.
+    In the mentoBucketsAfter updated the referenced logic changes to:
+
+    function getUpdatedBuckets(PoolExchange memory exchange) internal view returns (uint256 bucket0, uint256 bucket1) {
+      bucket0 = exchange.config.stablePoolResetSize;
+      uint256 exchangeRateNumerator;
+      uint256 exchangeRateDenominator;
+      (exchangeRateNumerator, exchangeRateDenominator) = getOracleExchangeRate(exchange.config.referenceRateFeedID);
+  
+      bucket1 = exchangeRateDenominator.mul(bucket0).div(exchangeRateNumerator);
+    }
+    */
+    const bucket0 = new BigNumber(
+      this.poolExchange.config.stablePoolResetSize._hex
+    );
+    const [exchangeRateNumerator, exchangeRateDenominator] =
+      await this.getOracleExchangeRate(
+        this.poolExchange.config.referenceRateFeedID
+      );
+    const bucket1 = exchangeRateDenominator
+      .multipliedBy(new BigNumber(bucket0))
+      .dividedBy(exchangeRateNumerator);
+    return { bucket0, bucket1 };
+  };
+
+  protected getOracleExchangeRate = async (
+    rateFeedID: Address
+  ): Promise<[BigNumber, BigNumber]> => {
+    const sortedOracles = ISortedOracles__factory.connect(
+      this.sortedOraclesAddress,
+      this.provider
+    );
+    const [rateNumerator, rateDenominator] = await sortedOracles.medianRate(
+      rateFeedID
+    );
+    if (rateDenominator.lte(0))
+      throw new Error("exchange rate denominator must be greater than 0");
+    return [
+      new BigNumber(rateNumerator._hex),
+      new BigNumber(rateDenominator._hex),
+    ];
+  };
+
+  private async getPricingModuleName(
+    address: Address
+  ): Promise<PricingFunctionType> {
+    const pricingModule = IPricingModule__factory.connect(
+      address,
+      this.provider
+    );
+    const name = await pricingModule.name();
+    if (name in PricingFunctionType) {
+      return name as unknown as PricingFunctionType;
```

Comment:
this typecasting seems a little weird. I think another way you can do this is something like this:
return PricingFunctionType[name as keyof typeof PricingFunctionType]

also a style nit here:
can you change this so that error case is checked first and main case doesn't need indentation:
i.e.
```
if (!(name in ...) { throw }

return PricingFunctionType[name as keyof typeof PricingFunctionType]
```